scenes/controller/offline_controller/offline_controller.gd

```diff
@@ -0,0 +1,162 @@
+class_name OfflineController
+extends Node
+
+@export var worker_controller: WorkerController
+@export var enemy_controller: EnemyController
+
+###############
+## overrides ##
+###############
+
+
+func _ready() -> void:
+	_connect_signals()
+
+
+#############
+## helpers ##
+#############
+
+
+## workers must be "happy" to work while not observed (none of the resources are decreasing)
+## TODO: how to implement "unhappy" workers working while not observed? (oscillating resources, ...)
+func _progress_worker_controller(efficiencies: Dictionary, cycles: int) -> Dictionary:
+	var total_efficiency: Dictionary = efficiencies["total_efficiency"]
+
+	# normalize "house -> worker" production
+	if total_efficiency.has("house"):
+		var worker_inc: int = Limits.safe_mult(
+			Game.params["house_workers"], total_efficiency.get("house", 0)
+		)
+		total_efficiency["worker"] = worker_inc
+	else:
+		total_efficiency["worker"] = 0
+
+	# check for decreasing resources
+	var ids: Array = total_efficiency.keys()
+	var decreasing_ids: Array = ids.filter(
+		func(id: String) -> bool: return total_efficiency[id] < 0
+	)
+	if !decreasing_ids.is_empty():
+		return {"decreasing_ids": decreasing_ids, "generated": {}}
+
+	# calculate production for cycles
+	var generated: Dictionary = {}
+	for id: String in ids:
+		var id_eff: int = total_efficiency[id]
+		var resource: ResourceGenerator = Resources.resource_generators[id]
+		if resource.is_dynamic_efficiency():
+			var de_id: String = resource.dynamic_efficiency_id
+			var de_id_eff: int = total_efficiency[de_id]
+			var generate: int = MathUtils.safe_dual_sum(id_eff, de_id_eff, cycles)
+			generated[id] = Limits.safe_add(generated.get(id, 0), generate)
+		else:
+			var generate: int = Limits.safe_mult(id_eff, cycles)
+			generated[id] = Limits.safe_add(generated.get(id, 0), generate)
+
+	return {"decreasing_ids": {}, "generated": generated}
+
+
+func _generate_resources(generated: Dictionary) -> bool:
+	var is_success: bool = false
+	for resource_id: String in generated:
+		var amount: int = generated[resource_id]
+		if amount == 0:
+			continue
+		is_success = true
+		SignalBus.resource_generated.emit(resource_id, amount, name)
+
+		var resource: ResourceGenerator = Resources.resource_generators[resource_id]
+		var worker_costs: Dictionary = resource.worker_costs
+		for worker_id: String in worker_costs:
+			var cost: int = Limits.safe_mult(amount, worker_costs[worker_id])
+			SignalBus.worker_generated.emit(worker_id, -cost, name)
+	return is_success
+
+
+func _progress_enemy_controller(
```

Comment:
Does fighting progress while offline? 

If I start on rabbit and go afk for a week would I come back to being on the last boss?

Reply:
Currently no, because you have to choose to either spare or kill a defeated enemy. 

(Overkill damage does not translate to next enemy.)